UsageEnvironment/Log.cpp

```diff
@@ -0,0 +1,352 @@
+
+  if (!backupLogFile.is_open()) {
+    perror("Couldn't open backup log file.");
+    return;
+  }
+
+  unsigned int originalLogNum = logFileNum;
+
+  NextLogFile();
+  if (access(currLogFilePath.c_str(), F_OK) != 0) {
+    // No wrap-around has ocurred yet
+    SetLogFileNum(0);  // Start from log 0
+  }
+
+  // Copy all logs into backup log file
+  for (int i = 0; i < kMaxLogs; i++) {
+    if (access(currLogFilePath.c_str(), F_OK) == 0) {
+      ifstream tmpLogFile(currLogFilePath);
+
+      if (!tmpLogFile.is_open()) {
+        perror("Couldn't open temp log file.");
+        break;
+      }
+
+      // Read temp log file contents
+      string line;
+      while (getline(tmpLogFile, line)) {
+        backupLogFile << line << endl;
+      }
+
+      // Add new line to end of file contents to prevent line-chaining
+      backupLogFile << '\n';
+
+      tmpLogFile.close();
+      NextLogFile();
+    }
+  }
+
+  SetLogFileNum(originalLogNum);
+
+  // Give everyone rw permissions for this backup file
+  mode_t filePerms = S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
+  if (chmod(backupFilePath, filePerms) != 0) {
+    perror("Error setting file permissions");
+  }
+
+  // Flush backup and close it
+  backupLogFile.flush();
+  backupLogFile.close();
+
+  // Flush all stdout and stderr
+  fflush(stderr);
+  fflush(stdout);
+}
+
+void _Log::ReleaseFile(FILE* logFileHandle) {
+  if (logFileHandle != NULL) {
+    fflush(logFileHandle);
+    struct stat file_stat;
+    if (currLogFilePath != "") {
+      // The file is not stdout or stderr, so we can back it up if needed
+      // then close it
+      if (stat(currLogFilePath.c_str(), &file_stat) < 0) {
+        perror("Error getting live555 log file stat.");
+      }
+      fflush(stdout);
+      fflush(stderr);
+      fclose(logFileHandle);
+#ifdef LOG_FILE_TMP_DIR
+      if (file_stat.st_size >= (MAX_LOG_FILE_SIZE / kMaxLogs)) {
+        PrepNextLog();
+      }
+#endif
+    }
+  }
+}
+
+void _Log::SetLevel(LogLevel level) {
+  logLevel = level;
+}
+
+/***
+ * Logging functions that take in file path and line number
+***/
+
+static string FormatLog(string fileName, int line, LogLevel logLevel, const char * format) {
+
+  string _format = string(timestamp()) + string(" ") + string(format);
+  _format += " [" + string(fileName) + string(":") + itoa(line) + "]";
+  return LogLevelAsString(logLevel) + string(" ") + string(_format) + "\n";
+}
+
+void _Log::Panic(const char* filePath, int line, const char * format, ...) {
+  lock_guard<mutex> logLockGuard(logMutex);
+  if (logLevel >= LogLevelPanic ) {
+    va_list argptr; va_start(argptr, format);
+    FILE* fileHandle = AcquireFile();
+    if (fileHandle == NULL) {
+      return;
+    }
+    vfprintf(fileHandle, FormatLog(filePath, line, LogLevelPanic, format).c_str(), argptr);
+    ReleaseFile(fileHandle);
+    va_end(argptr);
+  }
+}
+
+void _Log::Fatal(const char* filePath, int line, const char * format, ...) {
```

Comment:
I feel these are all redundant functions.
We can probably re-org and get away with 1 func + some args to indicate log level.

Reply:
Yousef is following Adeel's original design pattern. I agree it could be cleaned up, but not sure if it's worth it for this PR?

Reply:
yeah am okay if this taken up in a different PR.

Reply:
Cleaned them up.